### PR TITLE
Including an API that allows the GameController to be remote-controlled from an AutoReferee

### PR DESCRIPTION
**Pull request by MaikePaetzel**
This PR builds an additional GameControllerSimulator.jar that allows an automatic referee system to connect to the GameController via TCP/IP and send update control messages to the GameController. The API is properly documented in AutoRefSim.md 

Note that this is the first stable version of the API implementation. Future changes will occur to expand the API and make it more compatible with the simulated rules.

### CODE REVIEW

**Comment by omichel** [comment]
When the referee sends an answer to the GameController over the UDP port, it should provide a [team number and player number](https://github.com/RoboCup-Humanoid-TC/GameController/blob/master/protocols/python/receiver.py#L121-L122). Can I assume I shall pass 1 and 1 there as in the example? Hopefully, this won't conflict with actual player 1 of team 1. Or shall I pass special values? If yes, which ones?

**Comment by MaikePaetzel** [comment]
> 
> 
> When the referee sends an answer to the GameController over the UDP port, it should provide a [team number and player number](https://github.com/RoboCup-Humanoid-TC/GameController/blob/master/protocols/python/receiver.py#L121-L122). Can I assume I shall pass 1 and 1 there as in the example? Hopefully, this won't conflict with actual player 1 of team 1. Or shall I pass special values? If yes, which ones?

Yes, 1 and 1 works and it should not conflict with the actual players.

**Comment by omichel** [comment]
> Yes, 1 and 1 works and it should not conflict with the actual players.

Or maybe I should simply not send anything to the GameController over UDP, as it doesn't seems useful: the GameController doesn't seem to care about the referee heartbeat...

**Comment by MaikePaetzel** [comment]
> 
> 
> > Yes, 1 and 1 works and it should not conflict with the actual players.
> 
> Or maybe I should simply not send anything to the GameController over UDP, as it doesn't seems useful: the GameController doesn't seem to care about the referee heartbeat...

Ah right, this is how the heartbeat was implemented. I will need to think this through a little more, especially when it comes to the goal keeper communication as well. I'll get back to you about this

File: AutoRefAPI.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+### 4. Testing
+
+To test the TCP/IP message interface, you can run the ``main.py`` script
+in ``sim_client`` using Python 3.
+
+## TCP/IP Connection
+
+To connect to the GameController via the AutoReferee, you need to
+open a TCP/IP connection on port 8750. The port can be configured in
+the `AutoRefServer.java` class.
```
Comment (sheepsy90) [inline AutoRefAPI.md:61]: Maybe add this to the config file.

File: AutoRefAPI.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+require a different format!
+
+### Score
+
+Format: ``<id>:SCORE:<team_id>``
+
+This message indicates that a goal was scored by a given team.
+The ``team_id`` is the unique identifier of a team that needs to match the ID indicated in the
+``game.json`` config.
+
```
Comment (sheepsy90) [inline AutoRefAPI.md:163]: Maybe add an example trace of a couple commands to eg. start a game

File: sim_client/main.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
+
+import socket
+import time
+
+HOST = '127.0.0.1'  # The server's hostname or IP address
```
Comment (sheepsy90) [inline sim_client/main.py:6]: Maybe test / allow 0.0.0.0 so another machine in the network could connect to the GC

Comment (sheepsy90) [reply sim_client/main.py:6]: Oh thats the python script - nevermind

File: src/controller/net/AutoRefServer.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            boolean isConnected;
+            try {
+                if (simulator_socket != null && connectionSocket != null && simulator_socket.isBound()) {
+                    connectionSocket.close();
+                    simulator_socket.close();
+                }
+                simulator_socket = new ServerSocket(port);
+                this.connectionSocket = simulator_socket.accept();
+                isConnected = true;
+                System.out.println("Client Got Connected  ");
```
Comment (sheepsy90) [inline src/controller/net/AutoRefServer.java:42]: MAybe put those print statements behind a debug flag so they can be turned on/off

**Review by sheepsy90 — APPROVED** [review]
Looks good! Great work!

File: AutoRefAPI.md (hunk 1 above)
Comment (MaikePaetzel) [reply AutoRefAPI.md:61]: I actually don't think this makes sense in the config file but I will add a launch parameter for this case

File: AutoRefAPI.md (hunk 2 above)
Comment (MaikePaetzel) [reply AutoRefAPI.md:163]: Sample to start a game is already provided in the sim_client/main.py -> I added this to the documentation

File: src/controller/net/AutoRefServer.java (hunk 4 above)
Comment (MaikePaetzel) [reply src/controller/net/AutoRefServer.java:42]: Future work.

**Pull request merged**